### PR TITLE
introduce low memory flag

### PR DESCRIPTION
**Pull request by whyrusleeping**
I want to try an experiment. I want to have an ipfs 'low memory mode' that when enabled, limits ipfs's memory usage by turning down various numbers, reduces concurrency, and makes in memory caches smaller.

If anyone has any thoughts on where these changes should be made, feel free to post proposals.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

**Comment by whyrusleeping** [comment]
My idea here is that we find a bunch of places where we can tweak things to make usage lower, and make a 'binary' switch for them now. Later, once we've matured the option a bit, we can make various 'levels' of memory consumption mode (maybe even a high memory mode) that would affect the same areas of code.

**Comment by rht** [comment]
A global memory limit was proposed in #1482, which won't be enabled until go1.6 (https://github.com/golang/go/issues/10460).

**Comment by lgierth** [comment]
This will be great for embedded devices!

**Comment by rht** [comment]
Never mind embedded devices, even [VPS](https://github.com/ipfs/go-ipfs/issues/1482#issuecomment-154778400) got choked with the ram consumption.

Though there is always ulimit (when in unix).

**Comment by whyrusleeping** [comment]
right, the point of this PR wont be to set a hard number limit on memory, it will be policy changes to attempt to use less memory.

**Comment by rht** [comment]
[concurrentFdDials](https://github.com/ipfs/go-ipfs/blob/20b06a4cbce8884f5b194da6e98cb11f2c77f166/p2p/net/swarm/swarm_dial.go#L44)?

**Comment by whyrusleeping** [comment]
@rht ooooh, yeah. thats a good one

File: p2p/net/swarm/swarm_dial.go [hunk 1]
```diff
@@ -41,7 +42,12 @@ var (
 const dialAttempts = 1
 
 // number of concurrent outbound dials over transports that consume file descriptors
-const concurrentFdDials = 160
+var concurrentFdDials = 160
+
+func init() {
+	concurrentFdDials = 80
```
Comment (rht) [inline p2p/net/swarm/swarm_dial.go]: you forgot `if u.LowMemMode`

Comment (whyrusleeping) [inline p2p/net/swarm/swarm_dial.go]: *facepalm* I'm good at this...

File: cmd/ipfs/main.go [hunk 2]
```diff
@@ -182,6 +182,9 @@ func (i *cmdInvocation) Run(ctx context.Context) (output io.Reader, err error) {
 	if u.GetenvBool("DEBUG") {
 		u.Debug = true
 	}
+	if u.GetenvBool("IPFS_LOW_MEM") {
+		u.LowMemMode = true
```
Comment (jbenet) [inline cmd/ipfs/main.go]: maybe this should go into the config? ok to start like this i guess

**Comment by jbenet** [comment]
great idea! could also
- make it accept a memory bound and try to stick to that (back to resource constraints)
- run test suites with low mem on
- finally add that test with a max memory, and start hiking that lower

File: cmd/ipfs/main.go (hunk 2 above)
Comment (whyrusleeping) [inline cmd/ipfs/main.go]: Yeah, my thought was to keep this 'unofficial' until we're more confident about it working. Would suck to have a low memory mode that increased memory for some people, lol

**Comment by whyrusleeping** [comment]
> - make it accept a memory bound and try to stick to that (back to resource constraints)

I definitely want that, but I think it will be best to get this out there to get a better feel for its effect on various workloads first. Then once we have a better understanding, start being smarter

> - run test suites with low mem on

Yep! I'll enable that

> - finally add that test with a max memory, and start hiking that lower

mmmm, that would be nice. definitely on the todo list

**Comment by jbenet** [comment]
> Yep! I'll enable that

we should run the test suite over both cases. it may have effects in hiding or showing bugs.

**Pull request closed without merging**